plonk-core/src/proof_system/pi.rs

```diff
@@ -0,0 +1,193 @@
+//! inputs in evaluation or coefficient form.
+
+use alloc::collections::BTreeMap;
+use ark_ff::{FftField, ToConstraintField};
+use ark_poly::{
+    polynomial::UVPolynomial, univariate::DensePolynomial, EvaluationDomain,
+    GeneralEvaluationDomain,
+};
+use ark_serialize::{
+    CanonicalDeserialize, CanonicalSerialize, Read, SerializationError, Write,
+};
+
+use crate::prelude::Error;
+
+///  Public Inputs
+#[derive(
+    CanonicalDeserialize, CanonicalSerialize, Clone, PartialEq, Debug, Hash,
+)]
+pub struct PI<F>
+where
+    F: FftField,
+{
+    // padded size of the circuit
+    n: usize,
+    // non-zero values of the public input
+    values: BTreeMap<usize, F>,
+}
+
+impl<F> PI<F>
+where
+    F: FftField,
+{
+    /// Creates a new struct for [`PI`].
+    pub fn new(n: usize) -> Self {
+        assert!(n.is_power_of_two());
+        Self {
+            n,
+            values: BTreeMap::new(),
+        }
+    }
+
+    /// Updates the size of the circuit.
+    pub fn update_size(&mut self, n: usize) {
+        assert!(n.is_power_of_two());
+        self.n = n;
+    }
+
+    /// Inserts a new public input value at a given position.
+    ///
+    /// If the value provided is zero no insertion will be made as zeros are the
+    /// implicit value of empty positions.
+    /// The function will panic if an insertion is atempted in an already
+    /// occupied position.
+    pub fn insert(&mut self, pos: usize, val: F) {
+        if self.values.contains_key(&pos) {
+            panic!("Insertion in public inputs conflicts with previous value at position {}", pos);
+        }
+        if val != F::zero() {
+            self.values.insert(pos, val);
+        }
+    }
+
+    /// Inserts public input data that can be converted to one or more field
+    /// elements starting at a given position.
+    /// Returns the number of field elements occupied by the input or
+    /// [`Error::InvalidPublicInputValue`] if the input could not be converted.
+    pub fn add_input<T>(&mut self, pos: usize, item: &T) -> Result<usize, Error>
+    where
+        T: ToConstraintField<F>,
+    {
+        self.extend(pos, [item])
+    }
+
+    /// Inserts all the elements of `iter` converted into constraint field
+    /// elements in consecutive positions.
+    /// Returns the number of field elements occupied by `iter`
+    /// [`Error::InvalidPublicInputValue`] if the input could not be
+    /// converted.
+    fn extend<'t, T, I>(
+        &mut self,
+        init_pos: usize,
+        iter: I,
+    ) -> Result<usize, Error>
+    where
+        T: 't + ToConstraintField<F>,
+        I: IntoIterator<Item = &'t T>,
+    {
+        let mut result = 0;
+        for (pos1, item) in iter.into_iter().enumerate() {
+            let item_repr = &item
+                .to_field_elements()
+                .ok_or(Error::InvalidPublicInputValue)?;
+
+            for (pos2, elem) in item_repr.iter().enumerate() {
+                self.insert(init_pos + pos1 + pos2, *elem);
```

Comment:
I think you need to accumulate as you go, this won't insert in the right place. Take for example the following structure

```
[v0, [v1, v2], v3, v4]
```

In this case the loops would put them at the following indices:

```
v0 -> 0 + 0 = 0  (element 0 of group 0)
v1 -> 1 + 0 = 1  (element 0 of group 1)
v2 -> 1 + 1 = 2  (element 1 of group 1)
v3 -> 2 + 0 = 2  (element 0 of group 2)
v4 -> 3 + 0 = 3  (element 0 of group 3)
```

In this case, we have a conflict at index `3`. Instead, you can just loop over the arrays and count as you go:

```rust
let mut count = 0;
for item in iter {
    for elem in item.to_field_elements().ok_or(Error::InvalidPublicInputValue)? {
        self.insert(init_pos + count, elem);
        count += 1;
    }
}
Ok(count)
```

Reply:
Good point :+1: